### PR TITLE
implement suggestions

### PR DESCRIPTION
**Pull request by PaulBredl**
## Description of changes
Implement suggestion query

  
## How has this been tested?

Please describe the test strategy you followed. If not tested, please explain why.

- [x] automated unit test
- [x] automated integration test
- [ ] manual, exploratory test

    
## Checklist before requesting a review

- [ ] My code is easy to understand
- [ ] My code follows the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of this project
- [ ] My code fulfills all acceptance criteria
- [ ] The test coverage (line and branch) is reasonably high, especially on Service classes and other classes with complex logic
- [ ] I have made corresponding changes to the documentation or
  the [wiki/adr](https://github.com/IT-REX-Platform/wiki/tree/main/adr)
- [ ] I made no breaking changes in the database schema or if so, I will perform a database migration

## Checklist for reviewer

- The code is easy to understand
- The code follows
  the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of
  this project
- The code is tested or if not, the reason is documented or discussed
- The added and existing tests reasonably cover the code change
- The code has no breaking changes in the database schema or if so, the assignee is aware of it and
  will perform a database migration

### CODE REVIEW

File: src/main/java/de/unistuttgart/iste/gits/content_service/service/SuggestionService.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    public List<Suggestion> createSuggestions(List<UUID> chapterIds, UUID userId, int amount, List<SkillType> skillTypes) {
+        userProgressDataCache.clear();
+
+        List<Stage> availableStages = sectionService.getSectionsByChapterIds(chapterIds)
+                .stream()
+                .flatMap(Collection::stream)
+                .flatMap(section -> getAvailableStagesOfSection(section, userId).stream())
+                .toList();
+
+        return Stream.concat(
```
Comment (BlueWaves2) [inline src/main/java/de/unistuttgart/iste/gits/content_service/service/SuggestionService.java:51]: Doesn't this mean that there exists a case where only optional content is suggested, even if there is required content ready for repetition?

**Review by BlueWaves2 — DISMISSED** [review]
Over all looks good to me.
But I wonder if we want to really allow for optional content to be potentially be considered before required content in suggestions.

File: src/main/java/de/unistuttgart/iste/gits/content_service/service/SuggestionService.java (hunk 1 above)
Comment (PaulBredl) [reply src/main/java/de/unistuttgart/iste/gits/content_service/service/SuggestionService.java:51]: So you think, optional content should be prioritized lower or something? Good point, would have been nice to get this feedback in the concept pull request and not after implementing it 😅

I guess I need to rewrite most of it now...

Comment (BlueWaves2) [reply src/main/java/de/unistuttgart/iste/gits/content_service/service/SuggestionService.java:51]: I think optional content could be suggested if there is still room left in the return list. 
So, in other words, scan required content first. Then, if there is still room, fill up the remaining spots left in the return list with optional content that is due. 
What do you think of that approach?

Comment (PaulBredl) [reply src/main/java/de/unistuttgart/iste/gits/content_service/service/SuggestionService.java:64]: fixed it like you said, makes sense! :)

Comment (BlueWaves2) [reply src/main/java/de/unistuttgart/iste/gits/content_service/service/SuggestionService.java:64]: If I understand your new implementation correctly, you now always take all required content, sort & filter them by skilltypes and then repeat the same and put the optional content after the required content.
But doesn't this mean, we don't check if the content is even due for repetition yet?

Also, this means optional content would never be reachable if sample size is smaller than required content list size

Comment (PaulBredl) [reply src/main/java/de/unistuttgart/iste/gits/content_service/service/SuggestionService.java:64]:
```
.sorted(comparing((Content content) -> Duration.between(now(), getRelevantLearnDate(content, userId)).toDays())
```
line 81 or so, there is your check for the due date

Comment (PaulBredl) [reply src/main/java/de/unistuttgart/iste/gits/content_service/service/SuggestionService.java:64]: > Also, this means optional content would never be reachable if sample size is smaller than required content list size

I mean okay I could make an performance improvement that the optional part is only evaluated if necessary, but that is done by the java stream api automatically, as far as I know.
The optional contents wont be filtered and sorted, if the required contents are enough

**Review by BlueWaves2 — APPROVED** [review]

**Pull request merged**